### ISSUE TITLE
Failed - Create LXC container: LXD doesn't have a uid/gid allocation

### ISSUE THREAD

**Issue by raviram203** [HSBawa/icp-ce-on-linux-containers#17]
Ran the command: `sudo ./create-cluster.sh -es=dev --host=pc`. Creates `ubuntu 18.04 LTS amd64 (release) (20190306) | x86_64` but fails to create `bionic-image-for-icp-lvm`. 
```
==> bionic-image-for-icp-lvm: Creating container...
==> bionic-image-for-icp-lvm: Error creating container: LXD command error: Error: Failed container creation: Create LXC container: LXD doesn't have a uid/gid allocation. In this mode, only privileged containers are supported
==> bionic-image-for-icp-lvm: Unregistering and deleting deleting container...
==> bionic-image-for-icp-lvm: Error deleting container: LXD command error: Error: not found
Build 'bionic-image-for-icp-lvm' errored: Error creating container: LXD command error: Error: Failed container creation: Create LXC container: LXD doesn't have a uid/gid allocation. In this mode, only privileged containers are supported
```
Both `/etc/subuid` and `/etc/subgid` have the same entry: `ravi:100000:65536`. Have attached the console output.

Tried command `./create-cluster.sh -es=dev --host=pc` -- same results.

[create-cluster-2.log](https://github.com/HSBawa/icp-ce-on-linux-containers/files/2943787/create-cluster-2.log)

Please see attacehd `lxd-state-after-2.txt`. The existing storage pool `default` disappeared, but I am not sure when. Where does the `ubuntu 18.04 LTS amd64 (release) (20190306) | x86_64` exist? I cannot see it in storage pool `icpce`.

[lxd-state-after-2.txt](https://github.com/HSBawa/icp-ce-on-linux-containers/files/2943848/lxd-state-after-2.txt)

**Comment by raviram203** [comment]
Does `create-cluster.sh` install LXD from apt? I found LXD v3.0.3 and purged it, but `/usr/bin/lxc` is still there. I tried to get rid of it using dpkg, but have not been able. Any suggestions? I also purged snap LXD which I had installed manually. Wonder if the model is to perform all these operations on a fresh VM. I use a physical machine. Maybe a mismatch of scenario.

**Comment by HSBawa** [comment]
> Does `create-cluster.sh` install LXD from apt? I found LXD v3.0.3 and purged it, but `/usr/bin/lxc` is still there. I tried to get rid of it using dpkg, but have not been able. Any suggestions? I also purged snap LXD which I had installed manually. Wonder if the model is to perform all these operations on a fresh VM. I use a physical machine. Maybe a mismatch of scenario. Though, it did not override tools like `kubectl` and `terraform` saying they were already installed. But installed apt LXD even though snap LXD was installed.
> 
> I see that I can set `N` to avoid installing supporting tools such as terraform etc. Is there a flag to tell create-cluster.sh not to install LXD?
Good idea. I will add this option.

**Comment by HSBawa** [comment]
> Ran the command: `sudo ./create-cluster.sh -es=dev --host=pc`. Creates `ubuntu 18.04 LTS amd64 (release) (20190306) | x86_64` but fails to create `bionic-image-for-icp-lvm`.
> 
> ```
> ==> bionic-image-for-icp-lvm: Creating container...
> ==> bionic-image-for-icp-lvm: Error creating container: LXD command error: Error: Failed container creation: Create LXC container: LXD doesn't have a uid/gid allocation. In this mode, only privileged containers are supported
> ==> bionic-image-for-icp-lvm: Unregistering and deleting deleting container...
> ==> bionic-image-for-icp-lvm: Error deleting container: LXD command error: Error: not found
> Build 'bionic-image-for-icp-lvm' errored: Error creating container: LXD command error: Error: Failed container creation: Create LXC container: LXD doesn't have a uid/gid allocation. In this mode, only privileged containers are supported
> ```
> Both `/etc/subuid` and `/etc/subgid` have the same entry: `ravi:100000:65536`. Have attached the console output.
> 
> Tried command `./create-cluster.sh -es=dev --host=pc` -- same results.
> 
> [create-cluster-2.log](https://github.com/HSBawa/icp-ce-on-linux-containers/files/2943787/create-cluster-2.log)
> 
> Please see attacehd `lxd-state-after-2.txt`. The existing storage pool `default` disappeared, but I am not sure when. Where does the `ubuntu 18.04 LTS amd64 (release) (20190306) | x86_64` exist? I cannot see it in storage pool `icpce`.
> 
> [lxd-state-after-2.txt](https://github.com/HSBawa/icp-ce-on-linux-containers/files/2943848/lxd-state-after-2.txt)

Looks like I also saw this error. Are you running script as root? Please see following link and if you can validate uid/guid?
```
https://askubuntu.com/questions/1091729/failed-container-creation-lxd-doesnt-have-a-uid-gid-allocation
"... it was all a question of editing /etc/subuid and /etc/subgid and adding"
root:1000000:65536
```

**Comment by raviram203** [comment]
At first, I ran the script as `sudo`. Later, I ran as user `ravi` and the script asked for `sudo` password. Result was the same uid/gid error. 
I manually deleted `/usr/bin/lxc` hope that does not cause problems. I reinstalled snap LXD.
If I set `OVERWRITE_CLIS=Y`, `INSTALL_KUBECTL=y` and `KUBECTL_VERSION=v1.13.3` will the script replace `kubectl` with the specified version?

**Comment by HSBawa** [comment]
Interesting. Are you using Cosmic?

**Comment by raviram203** [comment]
No. Still on 18.04.2 bionic. I have become fond of LTS releases. Used to jump to `.10` in the old days.

**Comment by HSBawa** [comment]
Can you please try adding value "root:1000000:65536" and run as root and see if helps? I tried on my PCs and worked fine. I also tried on VMs (Fyre and AWS) ... didn't see any issue like this. I had seen this issue on Cosmic but then I moved back to Bionic for stability reasons.

**Comment by raviram203** [comment]
Let's step back. Should I have run the script as normal user the first time?

**Comment by HSBawa** [comment]
> Let's step back. Should I have run the script as normal user the first time?

No

**Comment by raviram203** [comment]
I used `sudo create-cluster.sh <options>` the first time. Did you have to add `root:1000000:65536` to uid/gid on 18.04 host? I will give it a shot the next time. Where should I add logic to bypass LXD installation?

**Comment by HSBawa** [comment]
I will wrap LXD apt call with flag. Usually, if LXD is already installed, install won't do nothing.

**Comment by raviram203** [comment]
script might be expecting apt LXD. I see logic to check `3.0.3`. looks like script will refuse to run with snap LXD. Should I switch to apt LXD?

**Comment by raviram203** [comment]
You are working late. I need to take a break. will take a look tomorrow. if you don't have time to bypass LXD installation, I will put the code in tomorrow. I understand the script (a little more) now. Thanks for the help.

**Comment by HSBawa** [comment]
@raviram203 I have updated code with INSTALL_LXD option. Also added additional updates, that I planned for today.  Let me know how it goes.
Btw, I checked my PC's subgid and subuid and I have not manually updated any of these files. I also checked settings on one of the cloud VM ... and they match. I am guessing LXD install issue:
```
lxd:100000:65536
root:100000:65536
[username redacted]:165536:65536
```

**Comment by HSBawa** [comment]
> script might be expecting apt LXD. I see logic to check `3.0.3`. looks like script will refuse to run with snap LXD. Should I switch to apt LXD? Seems like dev team has not updated apt LXD in a while. script installs apt LXD unconditionally in `initialize`.

"3.0.x" is last version APT will support. After that all SNAP. 
Also 3.0.3 is recommended for production use that makes me believe it is most stable version.
I would use following commands to clean up LXD and install again (CAUTION: This will remove all your existing configuration too):

## I had weird issues with snap lxd. Lot of times, I go socket error using lxc command,  after restart. I gave up as I was not sure what I was doing wrong.

snap remove lxd  

## also removes lxc i.e. client tools.
apt remove lxd lxd-client --purge -y   

apt install lxd lxd-client -y

**Comment by raviram203** [comment]
I had trouble with snap lxd and switched to apt lxd. But lxd dev team seemed to stop apt deliveries at 3.0.3, so I went back to snap. Built ICP several times using your previous approach. Am running your latest version with snap LXD 3.10. Has created the bionic image and all containers. But stalled now:

`module.container-master_worker.lxd_container.icp_ce_master: Still creating... (27m40s elapsed)`

Would be okay but `lxc list` shows "master" as stopped. Am I stuck? Just FYI.... Will close this thread.
```
lxc list
+--------------+---------+---------------------+------+------------+-----------+
|     NAME     |  STATE  |        IPV4         | IPV6 |    TYPE    | SNAPSHOTS |
+--------------+---------+---------------------+------+------------+-----------+
| dev-master-0 | STOPPED |                     |      | PERSISTENT |           |
+--------------+---------+---------------------+------+------------+-----------+
| dev-mgmt-0   | RUNNING | 10.50.50.141 (eth0) |      | PERSISTENT |           |
+--------------+---------+---------------------+------+------------+-----------+
| dev-proxy-0  | RUNNING | 10.50.50.121 (eth0) |      | PERSISTENT |           |
+--------------+---------+---------------------+------+------------+-----------+
| dev-worker-0 | RUNNING | 10.50.50.201 (eth0) |      | PERSISTENT |           |
+--------------+---------+---------------------+------+------------+-----------+
| dev-worker-1 | RUNNING | 10.50.50.202 (eth0) |      | PERSISTENT |           |
+--------------+---------+---------------------+------+------------+-----------+
```

**Comment by HSBawa** [comment]
@raviram203 I have seen this pattern myself using LXD 3.0.3. I believe it is a terraform lxd issue.  But good part is it happened only once.  I would simply cancel installation (Ctrl-C), destroy cluster (`./destroy_cluster.sh` ) and start the install process again.

Let me know if you see this issue repeatedly.

**Comment by raviram203** [comment]
Magnificent work. Bravo!
The new code runs faster. Very smooth -- all 16 CPU threads are equally busy. 1 hour 25 minutes on a real hard drive. Should be faster on SSD.
